idea-plugin/src/com/reason/ide/annotations/ErrorAnnotator.java

```diff
@@ -1,113 +1,229 @@
+    Ninja ninja = bucklescript.readNinjaBuild(contentRoot.get());
+    for (String devSource : config.getDevSources()) {
+      VirtualFile devFile = contentRoot.get().findFileByRelativePath(devSource);
+      if (devFile != null && FileUtil.isAncestor(devFile.getPath(), sourceFile.getPath(), true)) {
+        ninja.addInclude(devSource);
       }
     }
 
-    return result;
+    if (!BsPlatform.findBscExecutable(project, sourceFile).isPresent()) {
+      LOG.info("Unable to find bsc.exe");
+      return null;
+    }
+
+    // Creates a temporary file on disk with a copy of the current document.
+    // It'll be used by bsc for a temporary compilation
+    Path tempFilePath = Paths.get(m_compilationDirectory.getPath(), sourceFile.getName());
+    try {
+      Files.write(tempFilePath, psiFile.getText().getBytes());
+    } catch (IOException e) {
+      LOG.error("Failed to write to temporary file.", e);
+      return null;
+    }
+    LOG.debug("Wrote contents to temporary file.", tempFilePath);
+
+    File cmtFile = new File(m_compilationDirectory, sourceFile.getNameWithoutExtension() + ".cmt");
+
+    List<String> arguments = new ArrayList<>();
+    arguments.add("-bs-super-errors");
+    arguments.add("-color");
+    arguments.add("never");
+    arguments.addAll(ninja.getPkgFlags());
+    arguments.addAll(ninja.getBscFlags());
+    for (String ppxPath : ninja.getPpxIncludes()) {
+      arguments.add("-ppx");
+      arguments.add(ppxPath);
+    }
+    if (!namespace.isEmpty()) {
+      arguments.add("-bs-ns");
+      arguments.add(namespace);
+    }
+    if (jsxVersion != null) {
+      arguments.add("-bs-jsx");
+      arguments.add(jsxVersion);
+    }
+    for (String bscInclude : ninja.getIncludes()) {
+      arguments.add("-I");
+      arguments.add(bscInclude);
+    }
+    arguments.add("-o");
+    arguments.add(cmtFile.getAbsolutePath());
+    arguments.add("-bin-annot");
+    arguments.add(tempFilePath.toString());
+
+    return new InitialInfo(psiFile, cmtFile, editor, arguments);
   }
 
   @Nullable
   @Override
-  public Collection<BsbErrorAnnotation> doAnnotate(
-      @NotNull Collection<BsbErrorAnnotation> collectedInfo) {
-    return collectedInfo.isEmpty() ? null : collectedInfo;
+  public AnnotationResult doAnnotate(@Nullable InitialInfo initialInfo) {
+    if (initialInfo == null) {
+      LOG.warn("Unable to annotate file. Annotator not ready.");
+      return null;
+    }
+
+    PsiFile sourcePsiFile = Objects.requireNonNull(initialInfo.sourcePsiFile);
+    Project project = sourcePsiFile.getProject();
+    VirtualFile sourceFile = sourcePsiFile.getVirtualFile();
+
+    long compilationStartTime = System.currentTimeMillis();
+
+    BscProcess bscProcess = BscProcess.getInstance(project);
+    BscProcessListener bscListener = new BscProcessListener();
+
+    Integer exitCode = bscProcess.run(sourceFile, initialInfo.arguments, bscListener);
+    if (exitCode == null) {
```

Comment:
I would use `exitCode == null || exitCode != 0` here

Reply:
This one took me a minute to think through.

There are 3 states for exit codes:

1. Error code is null --> means something went wrong, an exceptional situation
2. Error code is 0 --> all good, re-analyze the PSI, update (clear) error annotations
3. Error code is non-null, non-zero --> compilation error, update error annotations

Reply:
ok but in case error code is different from 0, do we really want to handle 2 cases ? I think writing a log is the only option we have. so even if null/non-zero has different meaning, I don't think it matters.

Reply:
ok, I'll combine null and non-zero